inference/core/workflows/execution_engine/v1/dynamic_blocks/block_scaffolding.py

```diff
@@ -113,21 +144,52 @@ def _get_python_code_imports(python_code: PythonCode) -> str:
 
 
 def create_dynamic_module(
-    block_type_name: str, python_code: PythonCode, module_name: str
+    block_type_name: str,
+    python_code: PythonCode,
+    module_name: str,
+    workspace_id: Optional[str] = None,
 ) -> types.ModuleType:
     imports = _get_python_code_imports(python_code)
     code = python_code.run_function_code
     if python_code.init_function_code:
         code += "\n\n" + python_code.init_function_code
     code = imports + code
-    try:
+
+    # If using Modal and local execution is disabled, validate code remotely
+    if (
+        WORKFLOWS_CUSTOM_PYTHON_EXECUTION_MODE == "modal"
+        and not ALLOW_CUSTOM_PYTHON_EXECUTION_IN_WORKFLOWS
+    ):
+        # Validate code in Modal sandbox for security
+        from inference.core.workflows.execution_engine.v1.dynamic_blocks.modal_executor import (
+            validate_code_in_modal,
+        )
+
+        # Use anonymous workspace if not provided
+        validation_workspace = workspace_id or "anonymous"
+
+        # This will raise if validation fails
+        validate_code_in_modal(python_code, validation_workspace)
+
+        # Create a stub module for local reference
         dynamic_module = types.ModuleType(module_name)
-        exec(code, dynamic_module.__dict__)
+        # Add placeholder function
+        setattr(
+            dynamic_module, python_code.run_function_name, lambda *args, **kwargs: None
```

Comment:
@PawelPeczek-Roboflow would love to review this part together.  Probably good this way, but not entirely sure about implications of dynamic module being kind of a stub class here for purposes of execution engine / other validation.